### PR TITLE
Added feature directionality function + some changes to run_test and …

### PR DESCRIPTION
**Pull request by sneha-desai**
…k_fold_validation

### CODE REVIEW

File: vulcanai/utils.py [hunk 1]
```diff
@@ -419,3 +430,163 @@ def stitch_datasets(df_list, on, index_list=None):
...
     return merged_df
+
+
+def feature_directionality(features, data, val_x, val_y, dense_net, filename):
+    """
+    Will conduct tests to figure out directionality of features by finding all the unique feature values present in the
+    dataset and setting the feature for all examples to every unique value of that feature.
+
+    Args:
+        features: list of all features in the model
```
Comment (rfratila) [inline vulcanai/utils.py]: is features a redundant input if it can be found in data?

File: vulcanai/utils.py [hunk 2]
```diff
@@ -419,3 +430,163 @@ def stitch_datasets(df_list, on, index_list=None):
     merged_df.fillna(0, inplace=True)
 
     return merged_df
+
+
+def feature_directionality(features, data, val_x, val_y, dense_net, filename):
```
Comment (rfratila) [inline vulcanai/utils.py]: let's throw the dense_net (let's just call it network) at the front of the inputs.

File: vulcanai/utils.py [hunk 3]
```diff
@@ -419,3 +430,163 @@ def stitch_datasets(df_list, on, index_list=None):
...
+
+    Args:
+        features: list of all features in the model
+        data: dataframe with the data, features, y values
+        val_x: the features of the examples in the validation set
+        vay_y: the labels of the examples in the validation set
+        filename: the name of the file to save the test results to
+    """
+    test_df = None
+    for j in range(len(features)):
```
Comment (rfratila) [inline vulcanai/utils.py]: I think this can be sped up with just looking at the train and val features as a collective without the split inner loop (so we get from O(n^2) > O(n))

File: vulcanai/utils.py [hunk 4]
```diff
@@ -419,3 +430,163 @@ def stitch_datasets(df_list, on, index_list=None):
...
+        test_df: dataframe with aggregated test results
+        feature: the feature that is being variesd
+        value: value of the feature that is being tested
+    """
+
+    temp_df = pd.DataFrame(columns=["Feature", "Value", "TP", "FP", "TN", "FN",
+                                        "Accuracy", "Sensitivity", "Specificity", "PPV",
+                                        "NPV", "AUC"])
+
+    if network.num_classes is None or network.num_classes == 0:
```
Comment (rfratila) [inline vulcanai/utils.py]: Why not just use the run_test method instead of duplicating the code?

Comment (rfratila) [reply vulcanai/utils.py]: run_test already returns a dict object which you can aggregate running several times. See how it is used in k-fold for an example of this.

File: vulcanai/model_tests.py [hunk 5]
```diff
@@ -187,18 +189,18 @@ def run_test(network, test_x, test_y, figure_path='figures', plot=True):
     print ('Average AUC: : {:.4f}'.format(np.average(all_class_auc)))
     return {
         'accuracy': accuracy,
-        'macro_sensitivity': sens_macro,
-        'macro_specificity': spec_macro,
+        'sensitivity': sens[0],
```
[the remark this answers is not in the document]
Comment (sneha-desai) [reply vulcanai/model_tests.py]: Because the macro metrics are a weighted average of the all the classes which is always equal to the accuracy and doesn't distinguish between the classes

File: vulcanai/model_tests.py [hunk 6]
```diff
@@ -187,18 +189,18 @@ def run_test(network, test_x, test_y, figure_path='figures', plot=True):
...
+        'ppv': ppv[0],
+        'npv': npv[0],
+        'f1': f1[0],
+        'auc': np.average(all_class_auc)
     }
 
 
 def k_fold_validation(network, train_x, train_y, k=5, epochs=10,
-                      batch_ratio=1.0, plot=False):
+                      batch_ratio=1.0, plot=False, extra_x=None, extra_y=None):
```
[the remark this answers is not in the document]
Comment (sneha-desai) [reply vulcanai/model_tests.py]: I tried to make it a general formula where extra_x and extra_y are defaulted to None if nothing is passed in, but has the option to be added to the validation set if there is a class imbalance and undersampling has been done for training (like in my case)

File: vulcanai/model_tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sneha-desai) [reply vulcanai/model_tests.py]: The code I've written there is just temporary for a binary class case, it will need to be modified to fit the general n class case

File: vulcanai/net.py [hunk 7]
```diff
@@ -409,6 +409,13 @@ def mse_loss(self, prediction, y):
         print("Using Mean Squared error loss")
         return lasagne.objectives.squared_error(prediction, y).mean()
 
+    def sensitive_cost(self, prediction, y):
+        print("Using Sensitive Cost Function")
+        sum_fn = np.sum(np.logical_and(prediction[:, 0] == 0, y[:,0] == 1))
```
[the remark this answers is not in the document]
Comment (sneha-desai) [reply vulcanai/net.py]: Yeah I was only trying to penalize the FN, since those are errors that we want to make sure we reduce (classifying someone who has suicidal ideations as not having them). If we penalize the FP class too, wouldn't that just be the same thing that the cross entropy function does?

File: vulcanai/utils.py (hunk 3 above)
Comment (sneha-desai) [reply vulcanai/utils.py]: I have to do one test run for each unique value of every feature while all other feature values remain unchanged. I'm not able to see how it can be done with less complexity that what I have. Can you explain further?

File: vulcanai/utils.py (hunk 4 above)
Comment (sneha-desai) [reply vulcanai/utils.py]: Okay i'll look into it.

File: vulcanai/model_tests.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai/model_tests.py]: for specific cases like these, lets take advantage of **kwargs

File: vulcanai/net.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai/net.py]: cross entropy simply looks at the difference in probability predicted vs actual. it won't take into account FN and FP since those are more complex abstract concepts. you would need to outline it yourself to get that sort of control over the training criterion.

File: vulcanai/utils.py (hunk 1 above)
Comment (rfratila) [reply vulcanai/utils.py]: Great thanks!

File: vulcanai/utils.py (hunk 3 above)
Comment (rfratila) [reply vulcanai/utils.py]: oh I see sorry I misunderstood. in that case, we should be wary of features that have a lot of values so we can stratify it to not slow down the whole process too much.

File: vulcanai/net.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sneha-desai) [reply vulcanai/net.py]: I'll add FP into the mix and see how the model performs then.
